projects/ping_delays.kp/knowledge.md

```diff
@@ -0,0 +1,253 @@
+
+**`meta/Date`** - The time the Telemetry code in Firefox sent the ping to the server, according to the client's clock, expressed as a Date string conforming to [RFC 7231](https://tools.ietf.org/html/rfc7231#section-7.1.1.1).
+
+**`meta/Timestamp`** - The time the ping was received by the server, according to the server's
+clock, expressed in nanoseconds since the epoch.
+
+**`payload/info/subsessionLength`** - The length of time over which the ping was collecting data, according to the client's clock, expressed as a number of seconds. Working backwards from **`creationDate`** this gives us a subsession start time which allows us to determine Reporting Delay.
+
+
+```python
+subset = get_pings_properties(pings, ["application/channel",
+                                      "creationDate",
+                                      "meta/creationTimestamp",
+                                      "meta/Date",
+                                      "meta/Timestamp",
+                                      "payload/info/subsessionLength"])
+```
+
+```python
+p = subset.take(1)[0]
+```
+
+```python
+p
+```
+
+
+
+
+    {'application/channel': u'release',
+     'creationDate': u'2017-01-10T02:01:31.551Z',
+     'meta/Date': u'Tue, 10 Jan 2017 02:01:31 GMT',
+     'meta/Timestamp': 1484013691737682688L,
+     'meta/creationTimestamp': 1.484013691551e+18,
+     'payload/info/subsessionLength': 67}
+
+
+
+Quick normalization: ditch any ping that doesn't have a subsessionLength, creationTimestamp, or Timestamp:
+
+
+```python
+subset = subset.filter(lambda p:\
+                       p["payload/info/subsessionLength"] is not None\
+                       and p["meta/Timestamp"] is not None\
+                       and p["meta/creationTimestamp"] is not None)
+```
+We'll be plotting Cumulative Distribution Functions today.
+
+
+```python
+MAX_DELAY_S = 60 * 60 * 96.0
+HOUR_IN_S = 60 * 60.0
+CHANNELS = ['release', 'beta', 'aurora', 'nightly']
+```
+
+```python
+def setup_plot(title, max_x):
+    plt.title(title)
+    plt.xlabel("Delay (hours)")
+    plt.ylabel("% of pings")
+
+    plt.xticks(range(0, int(max_x) + 1, 2))
+    plt.yticks(map(lambda y: y / 10.0, range(0, 11, 1)))
+
+    plt.ylim(0.0, 1.0)
+    plt.xlim(0.0, max_x)
+
+    plt.grid(True)
+
+def plot_cdf(data):
+    sortd = np.sort(data)
+    ys = np.arange(len(sortd))/float(len(sortd))
+
+    plt.plot(sortd, ys)
+```
+
+```python
+def calculate_delays(p):
+    reporting_delay = p["payload/info/subsessionLength"]
+    
+    created = datetime.fromtimestamp(p["meta/creationTimestamp"] / 1000.0 / 1000.0 / 1000.0)
+    received = datetime.fromtimestamp(p["meta/Timestamp"] / 1000.0 / 1000.0 / 1000.0)
+    sent = datetime.fromtimestamp(mktime_tz(parsedate_tz(p["meta/Date"]))) if p["meta/Date"] is not None else received
```

Comment:
Any time we get a ping from a build that doesn't have bug 1144778 (landed in 48). Clock skew is generally pretty small (under 1s most of the time. I would point to a reference but the last time I analyzed Date vs Timestamp I sadly didn't save the notebook), so I was okay with handling pre-48 numbers without adjusting for skew.